core/src/desul/atomics/SYCL.hpp

```diff
@@ -17,17 +17,69 @@ SPDX-License-Identifier: (BSD-3-Clause)
 
 namespace desul {
 
+// FIXME_SYCL We need to either use generic_space or figure out how to check for the
+// correct adress space in a SYCL-portable way.
+#if defined(__SYCL_DEVICE_ONLY__) && !defined(__NVPTX__)
 #define DESUL_IMPL_SYCL_ATOMIC_FETCH_OPER(OPER, TYPE)                              \
   template <class MemoryOrder>                                                     \
   TYPE atomic_fetch_##OPER(TYPE* dest, TYPE val, MemoryOrder, MemoryScopeDevice) { \
-    Impl::sycl_atomic_ref<TYPE, MemoryOrder, MemoryScopeDevice> dest_ref(*dest);   \
+    auto l = __SYCL_GenericCastToPtrExplicit_ToLocal<TYPE>(dest);                  \
+    if (l) {                                                                       \
+      Impl::sycl_atomic_ref<TYPE,                                                  \
+                            MemoryOrder,                                           \
+                            MemoryScopeDevice,                                     \
+                            sycl::access::address_space::local_space>              \
+          dest_ref(*dest);                                                         \
+      return dest_ref.fetch_##OPER(val);                                           \
+    } else {                                                                       \
```

Comment:
Why is it not
```C++
auto g = __SYCL_GenericCastToPtrExplicit_ToGlobal<TYPE>(dest);
if (g) { ...
```

Reply:
Do you want both checks? 
For one check, I choose the space to test pretty arbitrarily. Do you have any arguments for rather checking for the global space?

Reply:
No I was just wondering whether we should check that the pointer is indeed global.

Reply:
I don't feel overly strong about this. As of know, there is https://github.com/intel/llvm/blob/a3e9aab6a25a9fc69e805f1a93f409375ce1f7d1/sycl/include/CL/sycl/atomic_ref.hpp#L42-L49 but also https://github.com/intel/llvm/blob/a3e9aab6a25a9fc69e805f1a93f409375ce1f7d1/sycl/include/CL/sycl/atomic_ref.hpp#L116-L120. All that is to say that really only `global_space` and `local_space` are supported (and we would get rid of all of this as soon as `generic_space` is available).
At this point, I slightly prefer not to have another check for every call SYCL atomics. If you (or others) prefer we can add the check and adopt https://github.com/intel/llvm/blob/a3e9aab6a25a9fc69e805f1a93f409375ce1f7d1/sycl/include/sycl/ext/oneapi/sub_group.hpp#L442-L443.